### ISSUE TITLE
command line tool additions/changes

### ISSUE THREAD

**Issue by asmagill** [Hammerspoon/hammerspoon#1356]
Since this was originally mentioned in a pull request, I thought I'd list it here as a feature request as a reminder or in case someone else wants to tackle it before I do:

From #1344:
... I have had a couple of ideas on some additions, though I've no immediate plans to implement them, just sort of a "maybe someday wish list":

1. if Hammerspoon isn't running, optionally prompt to start it, kind of like the "Install Command Line Tools" that occurs when you try to run clang but haven't install Xcode.
2. a persistent command history, maybe even tying it into the Console's history (which can be made persistent with some Lua code)
3. lua/Hammerspoon errors should be output through stderr rather then combined into the stdout stream (at least when using the `-c` option) with an appropriate return code for shell scripts to detect

**Comment by zzamboni** [comment]
Other wishes for `hs` from me:
- add a `-f` option that reads the code from a file
- allow passing command-line parameters to the executed code, which should get stored in the `arg` table as for a regular Lua script.

**Comment by szymonkaliski** [comment]
Would it be possible to have `hs` work as replacement for `hs.console`? I'd much prefer to run `hs` in my `tmux` split and see all the logs, than opening new window...

**Comment by asmagill** [comment]
@szymonkaliski just to clarify... you want a mode (perhaps a specific command line argument to `hs`) which causes output traditionally bound for the console to be sent to the command line tool's instance instead/in-addition?  A `tee` for the console?

**Comment by szymonkaliski** [comment]
Yeah exactly :)

**Comment by cmsj** [comment]
I like the idea of hs acting as a remote console, but I would go for sending it messages in addition to the built-in console

**Comment by asmagill** [comment]
To allow bidirectional communication between Hammerspoon and the command line tool which can be initiated by either end is a significant departure from the original... How important is backwards compatibility?

Specifically:
1. if the new version of the CLI attaches to the old version of `hs.ipc` should it work in a "legacy" mode? This is probably easy if desired.
2. if the old version of the CLI attaches to the new version of `hs.ipc` should it work in a "legacy" mode? A little more difficult, but should also be doable if desired.
3. anything printed in the old CLI instance is also printed to the console... should this continue to be a default for the new one?  This will have to be a toggle to allow console messages to be replicated to the CLI instance since having both on would create an infinite loop... should it be the default, though?
4. should raw mode and adding a custom handler via `hs.ipc.handler` continue to be supported?  This will be a major PITA, especially if currently existing alternative handlers are to be supported... to properly catch errors so that they can be piped to stderr instead of stdout, I had to change how the input is pre-parsed.  Supporting raw/custom handler is doable; supporting existing handlers, much harder.

The up side to this rewrite is that console mirroring will be doable, lua errors will change the return code when the tool is used with -c or stdin is piped in, and Hammerspoon will be able to create as many local and remote CFMessagePort objects as you like allowing the possibility of communication with other third party programs that do so (probably not a big boon, but at least it will now be possible).

I hope to have something up for testing in a day or so.

**Comment by latenitefilms** [comment]
In my personal opinion, I don't think backwards compatibility matters that much. I'd much rather Hammerspoon be constantly innovating and improving at a rapid pace, than worrying too much about making sure legacy scripts work fine.

**Comment by asmagill** [comment]
@zzamboni just noticed your additions; these should be doable in the new version.

Regarding arguments, should we also add an argument that specifies that the command line tool should stop inspecting the arguments and just blindly allow any arguments after the delimiter (I think `--` is the convention) so that your code can interpret them?

Which brings to mind supporting `#! /usr/local/bin/hs` in a script... have to look up how that's done while keeping arguments to the script versus arguments in the `#!` clause separate...  This will probably take longer, but I'll add it to the list as well.

**Comment by zzamboni** [comment]
@asmagill thanks! I agree being able to specify `--` to indicate "end of options" would be nice.

**Comment by asmagill** [comment]
New question: when in an interactive session via the command line, should "global" assignments to variables actually affect the global variables available from the Hammerspoon console or be unique to the cli instance?

e.g. if I'm in the cli and I type `a = 100`, should typing `a` into the actual console return 100?

I'm using the environment argument to `load` when parsing the input, and setting an __index metatable method so that the cli instance can "see" the global name space and access all of the builtins, modules, etc. but to make it truly seamless, I also have to add a __newindex metamethod... the question is should I?  If I don't, then backwards compatibility is broken.  Note that without the __newindex, you can still mess with the actual global variables like so `_G.a = -1`

I'm leaning towards not adding the __newindex metamethod, or at the very least requiring an additional argument to add it.  Thoughts?

**Comment by asmagill** [comment]
Where it gets dicey to document the difference, however, is that if you reference a table that is defined in the global space (say `hs`) and then do something like this in the cli instance: `hs.touch = 1`, then it is actually modify the global version because `hs` returns a table so the `touch` key is no longer being put into the local environment to the load function.

And, if the __newindex exists by default, then adding `getmetatable(_ENV).__newindex = nil` to your command line instance would go back to a semi-isolated environment.

Thoughts?

**Comment by cmsj** [comment]
I think the isolated environment would be more surprising than the other way around. I would certainly expect it to just be an interface into the main HS environment. What would be the reasons for the isolation?

**Comment by asmagill** [comment]
For the way we've been using the command line tool, you're probably right.

One of the new ways this can be used (well, shortly, still got a few bugs to work out) is as an interpreter for a shell script, i.e. in the first line of a file as `#! /usr/local/bin/hs`

For a shell script like this, keeping scopes separate would allow for a certain laziness in script writers not keeping things local.

At any rate, it's why I asked for comments and thoughts... I'm not strongly wedded to either, and I don't know that having to explain the intricacies of `_ENV` with a `_G` fallback is worth the effort.  I guess I'll leave the `__newindex` method in and maybe make note somewhere of the line I gave in my last comment that can provide some scope separation should you wish it.

**Comment by zzamboni** [comment]
I agree that it would be best to leave it as it is, with the CLI having access to the main HS environment. As it is now, I see it almost as "another console", and a way to interface with the running HS.

**Comment by asmagill** [comment]
Ok, for those interested, check out https://github.com/asmagill/hammerspoon_core/tree/master/ipc2.

It's very rough, has no man page, though the usage docs are present for the cli tool (and way too long and convoluted, going to need to clean that out at some point), but it's ready for some testing if you're interested.

Expand `ipc2-v0.1.tar.gz` in your config dir (usually ~/.hammerspoon/) and it will install as `hs.ipc2` so it won't interfere with the existing `hs.ipc`; expand `cli/hs-v0.1.tar.gz` wherever you want -- the install functions of `hs.ipc2` are not useful atm -- but somewhere where it won't interfere with your existing `hs` tool, and the CFMessagePort created has the name `hsCommandLine`... basically this shouldn't affect what's part of the core application unless you really try.

In a nutshell: use the `-C` argument to mirror the console to the cli instance. Use `-P` to print the instances output to the primary console (these are mutually exclusive and both are disabled by default), If an argument starts with ~, ./, or /, then it is treated as a file name and all arguments afterwards are passed in unparsed.  Supports `#! /.../hs` at the top of a text file which, if made executable, will execute the file in hammerspoon... see `test.hs` in the cli directory.  Lua errors go to stderr, regular output goes to stdout.  Autodetects if stdin is a pipe or stdout is redirected and disables color (unless you explicitly enable it with `-N`) automatically -- in fact `-s` is unnecessary now, but included for backwards compatibility, arguments passed in can be examined in `_cli.args` and `_cli._args` -- the later is all arguments passed into `hs` while the former will be the arguments after `--` or a file name so it can be used by a script to parse arguments passed specifically to it (see `test.hs` and it will be clearer).  Multiple `-c cmd` pairs can be specified and they will all be executed in order...

Legacy mode is mentioned in the docs, but does nothing at the moment; I've come up with a way that I think will allow the code to autodetect if you're connecting to or with the old ipc and/or cli tool, but that's not added in yet...

I'm really leaning towards dropping the whole raw/hs.ipc.handler thing... this version allows you to create your own remote ports, and the cli tool allows you to choose them with `-m` or you can write your own client entirely, so... if someone really wants their own handler to preparse or create their own environment, it can now be done in a cleaner manner.

Think that's it for the moment...

...nighty night bunny rabbit...

**Comment by chdiza** [comment]
> if Hammerspoon isn't running, optionally prompt to start it, kind of like the "Install Command Line Tools" that occurs when you try to run clang but haven't install Xcode.

Yes, yes, yes.  Indeed I'd go further and add a flag to the `hs` CLI tool which will forcibly start HS if it wasn't already running (so the user can skip the prompt altogether).

**Comment by cmsj** [comment]
@Asmagill nice!

**Comment by asmagill** [comment]
* the new cli tool checks every 2 seconds to see if the remote port is invalid; if it is it will retry up to 5 times, 2 seconds apart, to reconnect... this gives a 12 second window for Hammerspoon to reload and as long as `require("hs.ipc2")` is in your `init.lua` this should be fine.  May tweak numbers or make them setable on the command line if there is interest (this was added so `-C` doesn't require you to hit enter in the terminal window after a reload before it starts cloning the console again).

* the tool can now also talk in legacy mode to the original `hs.ipc` (try `./hs -m Hammerspoon` if you want to test it out) Can't go the other direction yet, but I think I know how to auto-detect this as well.

This leaves:
* persistent command history.  Should be easy to store in a file, though the examples I've seen so far would mean abandoning readline compatibility and just use editline directly... probably not an issue, though
  * not sure about syncing with actual console though
  * should the file name be selectable as an argument?
  * not sure about tab completion (I'd forgotten how used to that I'd become!)
* prompting to launch/relaunch Hammerspoon.  I go back and forth on this because it will make the command line invocation blocking until you respond... and what about different port names?  allow this only if `-m name` is not specified? Plus it isn't really that Hammerspoon isn't running - it's that `hs.ipc`/`hs.ipc2` isn't loaded.

**Comment by asmagill** [comment]
Regarding cloning the console... currently it doesn't output the results of what you type into the console input line directly -- this is because this is not output with `print` but rather appended directly   to the console's text field within MJConsoleWindowController.m.

I don't see this as an issue; if I'm typing in the console, I want the results in the console; if I'm typing in the cli tool, I want the results in the cli tool; printed output, however, is useful in both since that's how callbacks, hs.logger, etc. output to the console.

**Comment by chdiza** [comment]
I'm a little bit confused about the "cloning the console" stuff.

Currently, I only use the CLI tool with the `-c` flag to do things like activating my fake geektool, and reloading my fake geeklets.  For these purposes, I don't want any output from the CLI tool.  Upon first activation of my fake geektool I see:
```
-- Loading extension: task
-- Loading extension: timer
```
in the Terminal, and honestly it's kind of annoying.  I would've thought those messages would be shown only in the console (and only if the console is actually open).

Is there, or could there be, in the new CLI tool, a flag for quietness?  E.g., `hs -q -c load_my_fake_geeklets()`?  Such a flag would cause `hs` to not emit anything to the terminal at all except for special stuff like
```
error: can't access Hammerspoon; is it running with the ipc module loaded?
```
(If that's even still around by the time the tweaks are done.)

**Comment by asmagill** [comment]
tl;dr - yes, I can add a quiet argument so that the only output consists of errors or what is actually returned from the commands you issue with `-c msg`

The message you're referring to aren't because of console cloning... console cloning refers to any message that is output to the console through the `print` command but from an action that was not initiated in the terminal instance... for example if you type `t = hs.timer.doEvery(5, function() hs.printf("It's %s, do you know where your children are?", os.date()) end)` in the actual Hammerspoon console it will start printing the message every 5 seconds.  While that is running, if you go to Terminal and use the `hs` command to get an interactive session, with the old tool, you see none of the messages being added to the Hammerspoon console; but with the new tool, you could add a `-C` argument (note the capital C) and these messages would appear in the interactive terminal session as well.

The messages you're referring to are output by `hs._coresetup` when a module is autoloaded... the simplest fix for either version of the cli tool is to not rely on autoloading -- put an explicit `require("hs.task") ; require("hs.timer")` into your `load_my_fake_geeklets` function.  Autoloading outputs a message to the console via print (so this is directed to whichever console the command occurs on) for each module it has to autoload for you.

All that said, adding a `-q` argument is doable and will only allow errors or the final result to be output (e.g. with the new tool `hs -q -c "print([[ yay ]])" -c "return  4/2"` would return only `2` and `yay` would disappear into the ether.

**Comment by chdiza** [comment]
OK, thanks!

**Comment by chdiza** [comment]
> put an explicit require("hs.task") ; require("hs.timer") into your load_my_fake_geeklets function.

That doesn't actually work, for some reason.  I can put them in that function, or I can put them at the top of my init file(s), and those messages still appear in the terminal.

**Comment by chdiza** [comment]
Also, for some reason it's *only* task and timer that get output to the terminal, even though the console shows that "styledtext" is getting autoloaded too.

**Comment by asmagill** [comment]
Actually, my bad... I'm completely wrong above... (not sure what I was thinking when I wrote that, so again, sorry!)

To explain, pre-requiring them does prevent the auto-load feature from actually *loading* the module since it will be stored in the lua global `package.loaded`, but since `require` doesn't stick the module into the `hs` hierarchy, the autoload code still gets invoked to link the module from `package.loaded` into `hs`.

Ok, the "proper" way to do it is to load the modules you require into local variables and then use the local variable rather than `hs.blah` when trying to use it, for example:

~~~lua
local timer = require("hs.timer")
timer.doEvery(5, function() ... end) -- rather then `hs.timer.doEvery`
~~~

You will be happy to know that `-q` has already been added to the new `hs` so you'll have that option in the future, but I don't have an eta for when I'll actually move it into a pull request to formally replace the existing `hs.ipc` module... probably before the next release, but not quite yet...

@cmsj, this is the umpteenth time discussion about supressing the auto-load messages has come up... while I personally do use them to help debug some of my code, it's of minor use, so ...

Here are my thoughts on what we could do:
* add a setting like `MJConfigFile` which would allow people to suppress the output
* or, since `require` is already wrapped in _coresetup, add a line like `if not rawget(hs, "module") then rawset(hs, "module", result) end` (probably need some more code to make sure it was a top level "hs" module)... you'd still have to pre-require within a function/file, but at least the auto-load message could be suppressed by those who want to.

Thoughts?

**Comment by chdiza** [comment]
OK, thanks @asmagill, I'll think about doing that.

**Comment by asmagill** [comment]
Oh, I also meant to add these to the list of options:
* remove the messages all together
* use `hs.logger` and log the message at a level that isn't displayed by default

**Comment by latenitefilms** [comment]
> @cmsj, this is the umpteenth time discussion about supressing the auto-load messages has come up... while I personally do use them to help debug some of my code, it's of minor use, so ...

My suggestion would be to use `log.df()` for the auto-load message instead of `printf()`. Personally, I feel like `print()` should never be used by Hammerspoon core or Extensions, it should always be using `hs.logger`.

**Comment by asmagill** [comment]
tab completion added to cli tool... still in early testing, but looks promising

**Comment by cmsj** [comment]
I'm sure the autoloading messages came up again very recently, and everyone agreed we could suppress them with hs.logger :)

I haven't had much time recently, otherwise I would have done it already. I'd also like to say again, that I would also prefer us to remove the ability to disable autoloading :D

**Comment by asmagill** [comment]
Ok, I think the example at https://github.com/asmagill/hammerspoon_core/tree/master/ipc2 is almost ready for me to rename to `hs.ipc` and replace the current core module; I think all it lacks now is proper documentation and a review of the lua side support functions (for installing the cli tool, setting colors, etc.)

The new version has the following:
* prompt to launch Hammerspoon if it isn't running (or autolaunch if given `-A` -- see test2.hs in the `cli` folder)
* auto detect if Hammerspoon reloads/restarts and attempts to rebuild the connection when it occurs, not after you enter another command.
* /path/file support as well as shebang support
* console mirroring
* stderr is used for lua errors
* auto-detects when stdin and/or stdout are redirected and acts appropriately
* tab completion
* arguments are visible within the instance in `_cli._args` (all of them) and `_cli.args` (the unparsed ones) and arguments after `--` or a /file/path are passed in un-parsed (i.e. they can be arbitrary and are not checked for validity by the cli tool)
* the message port name can be specified to the tool -- this allows connecting to your own instances of `hs.ipc2.localPort`
* quiet mode (only errors and results are displayed) with `-q`
* legacy support if you accidentally connect the cli to the old `hs.ipc` or the old cli connects to the new `hs.ipc2` (with a logged message)... for people with a proper installation via `hs.ipc.cliInstall`, this shouldn't be an issue, but for those of us that might have an old one lying around somewhere or needing to test between versions, it works.  Plus it makes working with your own instances easier.  I'll need to document an example sometime -- doing so in legacy mode is significantly easier than fully supporting the new cli tool, but both approaches should be documented, just not sure where since it doesn't quite fit in the current doc scheme.
* persistent command history (not synchronized with Hammerspoon console, though... this becomes problematic because (a) the Hammerspoon console history isn't persistent without additional Lua code and (b) the readline wrapper to editline is only partial and doesn't support append/merge/duplicate detection in the history -- if you have multiple cli instances running, the last one to close overwrites the saved history of the previous instances... still it's better than no history saving)

And it should also be noted that creating your own handler in `hs.ipc.handler` will not be supported; not sure if this will be a problem for anyone (I've not heard from anyone on here that uses this feature, but that doesn't mean no one out there is since the support was put in by the original author), but if it is, the effect can be replicated by creating your own instance of `hs.ipc2.localPort` and attaching the handler to the new local port's callback -- as stated above, I'll need to document this at some point as well.

Am I missing anything?

**Comment by zzamboni** [comment]
@asmagill I only today got around to testing `hs.ipc2` and the new `hs`, but it's been working great so far. I'm not too familiar with the `hs.ipc.handler` interactions so I can't vouch for nothing having broken there, but it looks good to me. Thanks!

**Comment by chdiza** [comment]
@asmagill The "-q" flag, requested [here](https://github.com/Hammerspoon/hammerspoon/issues/1356#issuecomment-301246140), doesn't seem to actually suppress the notices of extension loading:
```
-- Loading extension: task
-- Loading extension: timer
```
I had been using your workaround of doing something like `task = require("hs.task")` and using `task` instead of `hs.task`, and it was working (on Sierra anyway).  With the new -q flag available, I reverted the workaround and starting using the -q flag on the command line.  The extension loading notices immediately returned.

**Comment by chdiza** [comment]
My immediately prior post seems to not apply anymore since the recent CLI fixes.  Thanks!
